File: apps/vsc-plugin/src/services/mcp-config-writer.service.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import * as vscode from 'vscode';
import { OutputChannel } from 'vscode';
import { IdeIntegrationService } from './ide-integration.service';

export interface McpEntry {
  id: string;
  server: { command: string; args: string[] };
  env: Record<string, string>;
}

export interface McpConfigureRequest {
  scope: string;
  mcps: McpEntry[];
  targetAgents?: string[];
}

export interface McpWriteResult {
  agent: string;
  file: string;
  status: 'written' | 'error';
  error?: string;
}

export interface ConfiguredMcpInfo {
  agent: string;
  mcpIds: string[];
  configFile: string;
}

interface AgentAdapter {
  agentName: string;
  serverKey: string;
  extensionIds: string[];
  globalConfigPath(): string;
  projectConfigPath(projectRoot: string): string;
}

const HOME = os.homedir();

const ADAPTERS: AgentAdapter[] = [
  {
    agentName: 'Cursor',
    serverKey: 'mcpServers',
    extensionIds: ['cursor.cursor'],
    globalConfigPath: () => path.join(HOME, '.cursor', 'mcp.json'),
    projectConfigPath: (root: string) => path.join(root, '.cursor', 'mcp.json'),
  },
  {
    agentName: 'Windsurf',
    serverKey: 'mcpServers',
    extensionIds: ['codeium.codeium', 'codeium.windsurf'],
    globalConfigPath: () => path.join(HOME, '.codeium', 'windsurf', 'mcp_config.json'),
    projectConfigPath: (root: string) => path.join(root, '.windsurf', 'mcp_config.json'),
  },
  {
    agentName: 'VS Code',
    serverKey: 'servers',
    extensionIds: ['github.copilot-chat'],
    globalConfigPath: () => {
      if (process.platform === 'darwin') {
        return path.join(HOME, 'Library', 'Application Support', 'Code', 'User', 'settings.json');
      } else if (process.platform === 'win32') {
        return path.join(process.env.APPDATA || HOME, 'Code', 'User', 'settings.json');
      }
      return path.join(HOME, '.config', 'Code', 'User', 'settings.json');
    },
    projectConfigPath: (root: string) => path.join(root, '.vscode', 'mcp.json'),
  },
  {
    agentName: 'Claude Desktop',
    serverKey: 'mcpServers',
    extensionIds: ['anthropic.claude-code', 'anthropics.claude', 'saoudrizwan.claude-dev'],
    globalConfigPath: () => {
      if (process.platform === 'darwin') {
        return path.join(HOME, 'Library', 'Application Support', 'Claude', 'claude_desktop_config.json');
      } else if (process.platform === 'win32') {
        return path.join(process.env.APPDATA || HOME, 'Claude', 'claude_desktop_config.json');
      }
      return path.join(HOME, '.config', 'claude', 'claude_desktop_config.json');
    },
    projectConfigPath: (root: string) => path.join(root, '.claude', 'mcp.json'),
  },
  {
    agentName: 'Claude Code',
    serverKey: 'mcpServers',
    extensionIds: ['anthropic.claude-code'],
    globalConfigPath: () => path.join(HOME, '.claude.json'),
    projectConfigPath: (root: string) => path.join(root, '.mcp.json'),
  },
];

export class McpConfigWriterService {
  private static instance: McpConfigWriterService;
  private log: OutputChannel;

  private constructor(log: OutputChannel) {
    this.log = log;
  }

  static initialize(log: OutputChannel): McpConfigWriterService {
    McpConfigWriterService.instance = new McpConfigWriterService(log);
    return McpConfigWriterService.instance;
  }

  static getInstance(): McpConfigWriterService {
    if (!McpConfigWriterService.instance) {
      throw new Error('McpConfigWriterService not initialized');
    }
    return McpConfigWriterService.instance;
  }

  configure(request: McpConfigureRequest): McpWriteResult[] {
    const results: McpWriteResult[] = [];
    const installedIds = this.getInstalledExtensionIds();
    const projectRoot = this.getProjectRoot();

    for (const adapter of ADAPTERS) {
      const isInstalled = adapter.extensionIds.some((id) => installedIds.has(id));
      if (!isInstalled) { continue; }
      if (request.targetAgents && !request.targetAgents.includes(adapter.agentName)) { continue; }

      const configPath = request.scope === 'project' && projectRoot
        ? adapter.projectConfigPath(projectRoot)
        : adapter.globalConfigPath();

      try {
        this.writeConfigFile(configPath, adapter.serverKey, request.mcps);
        results.push({ agent: adapter.agentName, file: configPath, status: 'written' });
        this.log.appendLine(`MCP config written for ${adapter.agentName}: ${configPath}`);
      } catch (e) {
        const errorMsg = e instanceof Error ? e.message : String(e);
        results.push({ agent: adapter.agentName, file: configPath, status: 'error', error: errorMsg });
        this.log.appendLine(`MCP config error for ${adapter.agentName}: ${errorMsg}`);
      }
    }

    return results;
  }

  getConfiguredMcps(): ConfiguredMcpInfo[] {
    const results: ConfiguredMcpInfo[] = [];
    const installedIds = this.getInstalledExtensionIds();
    const projectRoot = this.getProjectRoot();

    for (const adapter of ADAPTERS) {
      const isInstalled = adapter.extensionIds.some((id) => installedIds.has(id));
      if (!isInstalled) { continue; }

      const paths = [adapter.globalConfigPath()];
      if (projectRoot) {
        const projPath = adapter.projectConfigPath(projectRoot);
        if (projPath !== paths[0]) { paths.push(projPath); }
      }

      for (const configPath of paths) {
        try {
          if (!fs.existsSync(configPath)) { continue; }

          const json = JSON.parse(fs.readFileSync(configPath, 'utf-8'));
          const serverKey = json.mcpServers ? 'mcpServers' : json.servers ? 'servers' : null;
          if (!serverKey) { continue; }

          const servers = json[serverKey];
          if (!servers || typeof servers !== 'object') { continue; }

          const ids = Object.keys(servers);
          if (ids.length > 0) {
            results.push({
              agent: adapter.agentName,
              mcpIds: ids,
              configFile: configPath,
            });
          }
        } catch (e) {
          this.log.appendLine(`Failed to read config for ${adapter.agentName} at ${configPath}: ${e}`);
        }
      }
    }

    return results;
  }

  private writeConfigFile(configPath: string, serverKey: string, mcps: McpEntry[]): void {
    const dir = path.dirname(configPath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    let existing: Record<string, unknown> = {};
    if (fs.existsSync(configPath)) {
      try {
        existing = JSON.parse(fs.readFileSync(configPath, 'utf-8'));
      } catch {
        existing = {};
      }
    }

    const servers = (existing[serverKey] as Record<string, unknown>) || {};

    for (const mcp of mcps) {
      servers[mcp.id] = {
        command: mcp.server.command,
        args: mcp.server.args,
        env: mcp.env,
      };
    }

    existing[serverKey] = servers;
    fs.writeFileSync(configPath, JSON.stringify(existing, null, 2), 'utf-8');
  }

  private getInstalledExtensionIds(): Set<string> {
    const ids = new Set<string>();
    for (const ext of vscode.extensions.all) {
      ids.add(ext.id.toLowerCase());
    }
    return ids;
  }

  private getProjectRoot(): string | null {
    const folders = vscode.workspace.workspaceFolders;
    return folders && folders.length > 0 ? folders[0].uri.fsPath : null;
  }
}

File: apps/jetbrains-plugin/src/main/kotlin/com/windsurf/controller/services/McpConfigWriterService.kt
package com.windsurf.controller.services

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.ProjectManager
import java.io.File
import java.nio.file.Paths

data class McpEntry(
    val id: String,
    val server: McpServerDef,
    val env: Map<String, String>
)

data class McpServerDef(
    val command: String,
    val args: List<String>
)

data class McpConfigureRequest(
    val scope: String,
    val mcps: List<McpEntry>,
    val targetAgents: List<String>?
)

data class McpWriteResult(
    val agent: String,
    val file: String,
    val status: String,
    val error: String? = null
)

@Service(Service.Level.APP)
class McpConfigWriterService {

    private val logger = Logger.getInstance(McpConfigWriterService::class.java)
    private val gson: Gson = GsonBuilder().setPrettyPrinting().create()

    private val homeDir: String = System.getProperty("user.home")

    interface ConfigAdapter {
        val agentName: String
        val agentPluginIds: Set<String>
        fun globalConfigPath(): String
        fun projectConfigPath(projectRoot: String): String
        fun buildConfigJson(mcps: List<McpEntry>): String
    }

    // ── Cursor ─────────────────────────────────────────────
    private inner class CursorAdapter : ConfigAdapter {
        override val agentName = "Cursor"
        override val agentPluginIds = setOf("com.cursor.ide")

        override fun globalConfigPath(): String =
            Paths.get(homeDir, ".cursor", "mcp.json").toString()

        override fun projectConfigPath(projectRoot: String): String =
            Paths.get(projectRoot, ".cursor", "mcp.json").toString()

        override fun buildConfigJson(mcps: List<McpEntry>): String {
            val servers = JsonObject()
            for (mcp in mcps) {
                val entry = JsonObject()
                entry.addProperty("command", mcp.server.command)
                val argsArray = JsonArray()
                mcp.server.args.forEach { argsArray.add(resolveArg(it, mcp.env)) }
                entry.add("args", argsArray)
                val envObj = JsonObject()
                mcp.env.forEach { (k, v) -> envObj.addProperty(k, v) }
                if (envObj.size() > 0) entry.add("env", envObj)
                servers.add(mcp.id, entry)
            }
            val root = JsonObject()
            root.add("mcpServers", servers)
            return gson.toJson(root)
        }
    }

    // ── Windsurf / Codeium ─────────────────────────────────
    private inner class WindsurfAdapter : ConfigAdapter {
        override val agentName = "Windsurf"
        override val agentPluginIds = setOf("com.codeium.intellij")

        override fun globalConfigPath(): String =
            Paths.get(homeDir, ".codeium", "windsurf", "mcp_config.json").toString()

        override fun projectConfigPath(projectRoot: String): String =
            globalConfigPath()

        override fun buildConfigJson(mcps: List<McpEntry>): String {
            val servers = JsonObject()
            for (mcp in mcps) {
                val entry = JsonObject()
                entry.addProperty("command", mcp.server.command)
                val argsArray = JsonArray()
                mcp.server.args.forEach { argsArray.add(resolveArg(it, mcp.env)) }
                entry.add("args", argsArray)
                val envObj = JsonObject()
                mcp.env.forEach { (k, v) -> envObj.addProperty(k, v) }
                if (envObj.size() > 0) entry.add("env", envObj)
                servers.add(mcp.id, entry)
            }
            val root = JsonObject()
            root.add("mcpServers", servers)
            return gson.toJson(root)
        }
    }

    // ── VS Code + GitHub Copilot ───────────────────────────
    private inner class VSCodeAdapter : ConfigAdapter {
        override val agentName = "VS Code / GitHub Copilot"
        override val agentPluginIds = setOf("com.github.copilot")

        override fun globalConfigPath(): String =
            Paths.get(homeDir, ".vscode", "mcp.json").toString()

        override fun projectConfigPath(projectRoot: String): String =
            Paths.get(projectRoot, ".vscode", "mcp.json").toString()

        override fun buildConfigJson(mcps: List<McpEntry>): String {
            val servers = JsonObject()
            for (mcp in mcps) {
                val entry = JsonObject()
                entry.addProperty("type", "stdio")
                entry.addProperty("command", mcp.server.command)
                val argsArray = JsonArray()
                mcp.server.args.forEach { argsArray.add(resolveArg(it, mcp.env)) }
                entry.add("args", argsArray)
                val envObj = JsonObject()
                mcp.env.forEach { (k, v) -> envObj.addProperty(k, v) }
                if (envObj.size() > 0) entry.add("env", envObj)
                servers.add(mcp.id, entry)
            }
            val root = JsonObject()
            root.add("servers", servers)
            return gson.toJson(root)
        }
    }

    // ── Claude Code ────────────────────────────────────────
    private inner class ClaudeAdapter : ConfigAdapter {
        override val agentName = "Claude Code"
        override val agentPluginIds = setOf(
            "com.anthropic.claude", "anthropic.claude", "claude_code_terminal"
        )

        override fun globalConfigPath(): String =
            Paths.get(homeDir, ".claude.json").toString()

        override fun projectConfigPath(projectRoot: String): String =
            Paths.get(projectRoot, ".claude.json").toString()

        override fun buildConfigJson(mcps: List<McpEntry>): String {
            val servers = JsonObject()
            for (mcp in mcps) {
                val entry = JsonObject()
                entry.addProperty("command", mcp.server.command)
                val argsArray = JsonArray()
                mcp.server.args.forEach { argsArray.add(resolveArg(it, mcp.env)) }
                entry.add("args", argsArray)
                val envObj = JsonObject()
                mcp.env.forEach { (k, v) -> envObj.addProperty(k, v) }
                if (envObj.size() > 0) entry.add("env", envObj)
                servers.add(mcp.id, entry)
            }

            val existing = readExistingJson(globalConfigPath())
            existing.add("mcpServers", servers)
            return gson.toJson(existing)
        }
    }

    // ── JetBrains AI Assistant ──────────────────────────────
    private inner class JetBrainsAIAdapter : ConfigAdapter {
        override val agentName = "JetBrains AI Assistant"
        override val agentPluginIds = setOf("com.intellij.ai")

        override fun globalConfigPath(): String =
            Paths.get(homeDir, ".jb-mcp", "mcp.json").toString()

        override fun projectConfigPath(projectRoot: String): String =
            globalConfigPath()

        override fun buildConfigJson(mcps: List<McpEntry>): String {
            val servers = JsonObject()
            for (mcp in mcps) {
                val entry = JsonObject()
                entry.addProperty("command", mcp.server.command)
                val argsArray = JsonArray()
                mcp.server.args.forEach { argsArray.add(resolveArg(it, mcp.env)) }
                entry.add("args", argsArray)
                val envObj = JsonObject()
                mcp.env.forEach { (k, v) -> envObj.addProperty(k, v) }
                if (envObj.size() > 0) entry.add("env", envObj)
                servers.add(mcp.id, entry)
            }
            val root = JsonObject()
            root.add("mcpServers", servers)
            return gson.toJson(root)
        }
    }

    private val adapters: List<ConfigAdapter> = listOf(
        CursorAdapter(),
        WindsurfAdapter(),
        VSCodeAdapter(),
        ClaudeAdapter(),
        JetBrainsAIAdapter()
    )

    fun configure(request: McpConfigureRequest): List<McpWriteResult> {
        val results = mutableListOf<McpWriteResult>()
        val installedAgents = IdeIntegrationService.getInstance().detectInstalledAgents()
        val installedPluginIds = installedAgents.map { it.pluginId }.toSet() +
            installedAgents.map { it.id }.toSet()

        val projectRoot = getProjectRoot()

        for (adapter in adapters) {
            val isInstalled = adapter.agentPluginIds.any { it in installedPluginIds }
            if (!isInstalled) {
                logger.info("Skipping ${adapter.agentName}: not installed")
                continue
            }

            if (request.targetAgents != null) {
                val isTargeted = adapter.agentPluginIds.any { it in request.targetAgents }
                if (!isTargeted) {
                    logger.info("Skipping ${adapter.agentName}: not in targetAgents")
                    continue
                }
            }

            try {
                val configPath = when (request.scope) {
                    "project" -> {
                        if (projectRoot != null) adapter.projectConfigPath(projectRoot)
                        else adapter.globalConfigPath()
                    }
                    else -> adapter.globalConfigPath()
                }

                val configJson = adapter.buildConfigJson(request.mcps)
                writeConfigFile(configPath, configJson)

                results.add(McpWriteResult(
                    agent = adapter.agentName,
                    file = configPath,
                    status = "written"
                ))
                logger.info("Wrote MCP config for ${adapter.agentName}: $configPath")
            } catch (e: Exception) {
                results.add(McpWriteResult(
                    agent = adapter.agentName,
                    file = "",
                    status = "error",
                    error = e.message
                ))
                logger.error("Failed to write MCP config for ${adapter.agentName}", e)
            }
        }

        return results
    }

    private fun writeConfigFile(path: String, content: String) {
        val file = File(path)
        file.parentFile?.mkdirs()

        if (file.exists()) {
            val merged = mergeWithExisting(file, content)
            file.writeText(merged)
        } else {
            file.writeText(content)
        }
    }

    private fun mergeWithExisting(file: File, newContent: String): String {
        return try {
            val existing = gson.fromJson(file.readText(), JsonObject::class.java)
            val incoming = gson.fromJson(newContent, JsonObject::class.java)

            val serverKey = when {
                incoming.has("mcpServers") -> "mcpServers"
                incoming.has("servers") -> "servers"
                else -> return newContent
            }

            val existingServers = existing.getAsJsonObject(serverKey) ?: JsonObject()
            val newServers = incoming.getAsJsonObject(serverKey) ?: JsonObject()

            for (entry in newServers.entrySet()) {
                existingServers.add(entry.key, entry.value)
            }

            existing.add(serverKey, existingServers)
            gson.toJson(existing)
        } catch (e: Exception) {
            logger.warn("Could not merge with existing config, overwriting: ${e.message}")
            newContent
        }
    }

    private fun readExistingJson(path: String): JsonObject {
        return try {
            val file = File(path)
            if (file.exists()) {
                gson.fromJson(file.readText(), JsonObject::class.java) ?: JsonObject()
            } else {
                JsonObject()
            }
        } catch (e: Exception) {
            JsonObject()
        }
    }

    private fun resolveArg(arg: String, env: Map<String, String>): String {
        val regex = Regex("\\$\\{(\\w+)}")
        return regex.replace(arg) { match ->
            val key = match.groupValues[1]
            env[key] ?: match.value
        }
    }

    private fun getProjectRoot(): String? {
        return try {
            val projects = ProjectManager.getInstance().openProjects
            projects.firstOrNull()?.basePath
        } catch (e: Exception) {
            null
        }
    }

    data class ConfiguredMcpInfo(
        val agent: String,
        val mcpIds: List<String>,
        val configFile: String
    )

    fun getConfiguredMcps(): List<ConfiguredMcpInfo> {
        val results = mutableListOf<ConfiguredMcpInfo>()
        val installedAgents = IdeIntegrationService.getInstance().detectInstalledAgents()
        val installedPluginIds = installedAgents.map { it.pluginId }.toSet() +
            installedAgents.map { it.id }.toSet()
        val projectRoot = getProjectRoot()

        for (adapter in adapters) {
            val isInstalled = adapter.agentPluginIds.any { it in installedPluginIds }
            if (!isInstalled) continue

            val paths = mutableListOf(adapter.globalConfigPath())
            if (projectRoot != null) {
                val projPath = adapter.projectConfigPath(projectRoot)
                if (projPath != paths[0]) paths.add(projPath)
            }

            for (configPath in paths) {
                try {
                    val file = File(configPath)
                    if (!file.exists()) continue

                    val json = gson.fromJson(file.readText(), JsonObject::class.java) ?: continue
                    val serverKey = when {
                        json.has("mcpServers") -> "mcpServers"
                        json.has("servers") -> "servers"
                        else -> continue
                    }

                    val servers = json.getAsJsonObject(serverKey) ?: continue
                    val ids = servers.keySet().toList()
                    if (ids.isNotEmpty()) {
                        results.add(ConfiguredMcpInfo(
                            agent = adapter.agentName,
                            mcpIds = ids,
                            configFile = configPath
                        ))
                    }
                } catch (e: Exception) {
                    logger.warn("Failed to read config for ${adapter.agentName} at $configPath: ${e.message}")
                }
            }
        }

        return results
    }

    companion object {
        fun getInstance(): McpConfigWriterService =
            ApplicationManager.getApplication().getService(McpConfigWriterService::class.java)
    }
}
